### ISSUE TITLE
301 status code

### ISSUE THREAD

**Issue by echonox** [Rob--W/cors-anywhere#163]
I am trying to proxy this url on my self hosted server, locally it works fine but on my server i keep getting status code 301:

https://www.webcal.fi/cal.php?id=184&format=ics&locationname=Houtvester&lat=52.959289550781&lon=5.9185400009155&tempunit=c&speedunit=mps&color=%23C9DAF8&cntr=nl&lang=nl&rid=wc

This is the cors server options:

originBlacklist: originBlacklist,
  originWhitelist: originWhitelist,
  requireHeader: ['origin', 'x-requested-with'],
  checkRateLimit: checkRateLimit,
  removeHeaders: [
    'cookie',
    'cookie2',
    // Strip Heroku-specific headers
    'x-heroku-queue-wait-time',
    'x-heroku-queue-depth',
    'x-heroku-dynos-in-use',
    'x-request-start',
  ],
  redirectSameOrigin: true,
  httpProxyOptions: {
    // Do not add X-Forwarded-For, etc. headers, because Heroku already adds it.
    xfwd: false,
  },

**Comment by Rob--W** [comment]
CORS Anywhere should be hosted at its own domain. If you are sharing the domain for your application and CORS Anywhere, then you need to set `redirectSameOrigin` to `false`, to deactivate the logic at
https://github.com/Rob--W/cors-anywhere/blob/4814647a9d21f14858da282387a4295c17f2851d/lib/cors-anywhere.js#L356-L365

**Comment by echonox** [comment]
> CORS Anywhere should be hosted at its own domain. If you are sharing the domain for your application and CORS Anywhere, then you need to set `redirectSameOrigin` to `false`, to deactivate the logic at
> https://github.com/Rob--W/cors-anywhere/blob/4814647a9d21f14858da282387a4295c17f2851d/lib/cors-anywhere.js#L356-L365

Yes Cors Anywhere is hosted on its own domain

**Comment by Rob--W** [comment]
> Yes Cors Anywhere is hosted on its own domain

Separate from your main application?

Then the 301 redirect may actually be the real response from the proxied site.

**Comment by echonox** [comment]
> Separate from your main application?

Yes its separate from the main application but under the same server only at different port and endpoint so the main backend app is at mysite.com/b and the cors server endpoint is at mysite.com/node

> Then the 301 redirect may actually be the real response from the proxied site.

No when i run the link in browser or with curl it gives proper 200 response

**Comment by Rob--W** [comment]
>  Yes its separate from the main application but under the same server only at different port and endpoint so the main backend app is at mysite.com/b and the cors server endpoint is at mysite.com/node

So it is not separate. From the browser's perspective, it is a same-origin request. You need to update server.js and set `redirectSameOrigin`, as mentioned at https://github.com/Rob--W/cors-anywhere/issues/163#issuecomment-480756567 . And then clear the browser cache to forget about the permanent redirect (301 status code).

**Comment by echonox** [comment]
@Rob--W I tried setting redirectSameOrigin: false but still not working,

I noticed this, when I use the  http://127.0.0.1:333/url-to-fetch

the request is successful with 200 response

but when I use:
app.mysite.com/node/url-to-fetch I get response 301

and this response:

"Not found because of proxy error: Error: getaddrinfo ENOTFOUND www.app.mysite.com www.app.mysite.com:80"

Note in order to be able to do a call to the nodejs server from my apache2 server I use this configuration in the Virtual Host:

ProxyRequests on
ProxyPass /node/ http://127.0.0.1:3333/

With this modules enabled:

proxy, proxy_http, proxy_balancer, lbmethod_byrequests

this allows me to access 127.0.0.1:3333 from app.mysite.com/node/

**Comment by Rob--W** [comment]
You cannot have a 301 and "Not found because of proxy error:" at the same time. Those are two separate requests, first 301, and then another request (because of the redirect) to a response with error 404.

You should look at the exact request and response bodies for all requests, e.g. via the DevTools of your browser, or via a network sniffer such as Wireshark.

**Comment by echonox** [comment]
@Rob--W Added some console logs to the code:

> cors-anywhere@0.4.1 start ~/Dekstop/cors
> HOST=127.0.0.1 PORT=3333 CORSANYWHERE_WHITELIST=http://app.localsite.nl node server.js

Running CORS Anywhere on 127.0.0.1:3333
location variable:
 Url {
  protocol: 'http:',
  slashes: true,
  auth: null,
  host: 'www.webcal.fi',
  port: null,
  hostname: 'www.webcal.fi',
  hash: null,
  search:
   '?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%23C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573',
  query:
   'id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%23C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573',
  pathname: '/cal.php',
  path:
   '/cal.php?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%23C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573',
  href:
   'http://www.webcal.fi/cal.php?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%23C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573' }


statusCode:  301 
locationHeader:  http://www.app.klenderlocal.nl/cal.php?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%2523C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573 
requestState.location.href:  http://www.webcal.fi/cal.php?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%23C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573

location Url {
  protocol: 'http:',
  slashes: true,
  auth: null,
  host: 'www.app.localsite.nl',
  port: null,
  hostname: 'www.app.localsite.nl',
  hash: null,
  search:
   '?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%2523C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573',
  query:
   'id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%2523C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573',
  pathname: '/cal.php',
  path:
   '/cal.php?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%2523C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573',
  href:
   'http://www.app.localsite.nl/cal.php?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%2523C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573' }


So to clarify things:

The actual url that I want to proxy with the server and bypass cors is:
http://www.webcal.fi/cal.php?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%23C9DAF8&cntr=nl&lang=nl&rid=wc

The local apache server url is:
http://app.localsite.nl/

From the logs I see the cors server tries to access this url after getting error 301:

http://www.app.localsite.nl/cal.php?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%2523C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573

2 things wrong here, first the proxy adds www to the app.localsite.nl (this url doesn't exist)
and the second thing the "cal.php?id=184&format=ics&locationname=Utrechtse+Heuvelrug&lat=52.036336223378534&lon=5.435141563415527&tempunit=c&speedunit=mps&color=%2523C9DAF8&cntr=nl&lang=nl&rid=wc&_=1554900448573" belongs to the external url not the local server

**Comment by Rob--W** [comment]
Your log is confusing; it seems that you have partially redacted the content.

Anyway, the URL from the redirect seems to appear in `locationHeader`, i.e. the `Location` response header of the proxied request. This response is then forwarded with CORS headers to the original request.

From the perspective of the proxy server, everything seems to look all right.
